### PR TITLE
Use private instance of MemoryCache and impose size limit

### PR DESCRIPTION
**Pull request by JunTaoLuo**
Addresses #127 Still need to add some tests.

### CODE REVIEW

File: src/Microsoft.AspNetCore.ResponseCaching/Internal/Interfaces/IResponseCache.cs [hunk 1]
```diff
@@ -11,7 +11,7 @@ public interface IResponseCache
         IResponseCacheEntry Get(string key);
         Task<IResponseCacheEntry> GetAsync(string key);
 
-        void Set(string key, IResponseCacheEntry entry, TimeSpan validFor);
-        Task SetAsync(string key, IResponseCacheEntry entry, TimeSpan validFor);
+        void Set(string key, IResponseCacheEntry entry, TimeSpan validFor, long size);
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.ResponseCaching/Internal/Interfaces/IResponseCache.cs]: Looking into refactoring the helpers to the interface does not change.

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingMiddleware.cs [hunk 2]
```diff
@@ -30,6 +31,24 @@ public class ResponseCachingMiddleware
             IOptions<ResponseCachingOptions> options,
             ILoggerFactory loggerFactory,
             IResponseCachingPolicyProvider policyProvider,
+            IResponseCachingKeyProvider keyProvider)
+            : this(
+                next,
+                options,
+                loggerFactory,
+                policyProvider,
+                new MemoryResponseCache(new MemoryCache(new MemoryCacheOptions
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingMiddleware.cs]: Alternatively, we can add the MemoryResponseCache to DI but have it resolve ResponseCachingOptions to figure out the size.

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingOptions.cs [hunk 3]
```diff
@@ -9,6 +9,11 @@ namespace Microsoft.AspNetCore.ResponseCaching
     public class ResponseCachingOptions
     {
         /// <summary>
+        /// The size limit for the response cache middleware in bytes. The default is set to 100 MB.
+        /// </summary>
+        public long SizeLimit { get; set; } = 100 * 1024 * 1024;
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingOptions.cs]: Should we have ResponseCachingOptions extend MemoryCacheOptions so other options can be set?

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingMiddleware.cs [hunk 4]
```diff
@@ -505,5 +549,72 @@ internal static StringValues GetOrderCasingNormalizedStringValues(StringValues s
...
+                    foreach (var item in cachedResponse.Headers)
+                    {
+                        size += item.Key.Length * sizeof(char) + EstimateStringValuesSize(item.Value);
+                    }
+                }
+
+                // Body
+                if (cachedResponse.Body != null)
+                {
+                    size += cachedResponse.Body.Length * sizeof(char);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingMiddleware.cs]: bytes, not chars.

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingMiddleware.cs [hunk 5]
```diff
@@ -333,7 +368,16 @@ internal Task FinalizeCacheHeadersAsync(ResponseCachingContext context)
...
+
+                    try
+                    {
+                        await _cache.SetAsync(
+                            context.StorageVaryKey ?? context.BaseKey,
+                            context.CachedResponse,
+                            context.CachedResponseValidFor,
+                            EstimateCachedResponseSize(context.CachedResponse));
+                    }
+                    catch (OverflowException) { }
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingMiddleware.cs]: don't bother, we'll run out of RAM first.

**Comment by JunTaoLuo** [comment]
:up::date:

**Pull request merged**